resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceScheduler.java

```diff
@@ -0,0 +1,167 @@
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * 
+ * @author Nicolas NESMON
+ *
+ */
+class SseEventSourceScheduler
+{
+
+   private static class DaemonThreadFactory implements ThreadFactory
+   {
+
+      private final ThreadGroup group;
+
+      private final AtomicInteger threadNumber = new AtomicInteger(1);
+
+      private final String namePrefix;
+
+      DaemonThreadFactory(String name)
+      {
+         SecurityManager s = System.getSecurityManager();
+         group = (s != null) ? s.getThreadGroup() : Thread.currentThread().getThreadGroup();
+         namePrefix = name + "-thread-";
+      }
+
+      @Override
+      public Thread newThread(Runnable r)
+      {
+         Thread t = new Thread(group, r, namePrefix + threadNumber.getAndIncrement(), 0);
+         t.setDaemon(true);
+         return t;
+      }
+   }
+
+   private final ScheduledExecutorService scheduledExecutorService;
+
+   private final boolean shutdownExecutorService;
+
+   private final Phaser phaser;
+
+   private final AtomicBoolean closed;
+
+   SseEventSourceScheduler(ScheduledExecutorService scheduledExecutorService, String threadName, final Runnable onTerminated)
+   {
+      this.scheduledExecutorService = scheduledExecutorService == null
+            ? Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory(threadName))
+            : scheduledExecutorService;
+      this.shutdownExecutorService = scheduledExecutorService == null;
+      this.phaser = new Phaser(1)
+      {
+         @Override
+         protected boolean onAdvance(int phase, int registeredParties)
+         {
+            boolean terminated = super.onAdvance(phase, registeredParties);
+            if (terminated)
+            {
+               if (onTerminated != null)
+               {
+                  onTerminated.run();
+               }
+            }
+            return terminated;
+         }
+      };
+      this.closed = new AtomicBoolean(false);
+   }
+   
+   void schedule(final Runnable runnable, long delay, TimeUnit unit) throws RejectedExecutionException
+   {
+      if (this.closed.get())
+      {
+         return;
+      }
+      try
+      {
+         this.scheduledExecutorService.schedule(new Runnable()
+         {
+            @Override
+            public void run()
+            {
+               if (SseEventSourceScheduler.this.closed.get())
+               {
+                  return;
+               }
+               int registrationPhase = SseEventSourceScheduler.this.phaser.register();
+               try
+               {
+                  // If this phaser is terminated (registrationPhase < 0)
+                  // (no more registered parties) or the current
+                  // registration phase is other than 0 it means that
+                  // shutdownNow has been called already.
+                  if (registrationPhase != 0)
```

Comment:
@NicoNes  I  tried this locally and it turns out registrationPhase here can only be 0 before phaser is terminated because each job will be arriveAndDeregistered. 
In shutdownNow(),   phaser.arriveAndDeregister() can't guarantee it successfully terminates the phaser the scheduled job is running without call the phaser.arriveAndDeregiste(). 
```
 void shutdownNow()
   {
      if (this.closed.compareAndSet(false, true))
      {
         this.phaser.arriveAndDeregister();
         if (this.shutdownExecutorService)
         {
            this.scheduledExecutorService.shutdownNow();
         }
      }
   }
```
So IMO change this to phaser.forceTermination() and check phaser.isTerminated() here could be better and more guarantee. WDYT ?

Reply:
@NicoNes  sorry. I still didn't get what is the actual reason for the phase can only be 0. If I understand correctly, phaser.arriveAndDeregister() can't increase the arrived party number(or phaser.getPhase()) . So in your change, there is no code to increase the arrived party( or make the phaser.getPhase() larger than 0) , right ?
Regarding the without forceTermination,  I agree with you and it's better. 
For the onAdcance() method, if it is not called(phaser is not advanced), is it a problem without running the onTerminated runnable method ?
```
this.phaser = new Phaser(1)
      {
         @Override
         protected boolean onAdvance(int phase, int registeredParties)
         {
            boolean terminated = super.onAdvance(phase, registeredParties);
            if (terminated)
            {
               if (onTerminated != null)
               {
                  onTerminated.run();
               }
            }
            return terminated;
         }
      };
   ```

Reply:
Hey @jimma,

> I still didn't get what is the actual reason for the phase can only be 0. If I understand correctly, phaser.arriveAndDeregister() can't increase the arrived party number(or phaser.getPhase()) . So in your change, there is no code to increase the arrived party( or make the phaser.getPhase() larger than 0) , right ?

`phaser.arriveAndDeregister()` will increase the number of arrived parties for the phaser current phase and decrease the number of registered parties needed for future phases.
When all expected parties are arrived for phaser current phase, `onAdvance()` will be invoked. 
This method will say if the phaser is terminated or if it can advance to next phase.
The default `super.onAdvance()` will return true, indicating that the phaser is terminated if there is no more registered parties, else it will return false;

> For the onAdcance() method, if it is not called(phaser is not advanced), is it a problem without running the onTerminated runnable method ?

If `onAdvance()` is not called, it means that the phaser is not terminated and there are still running tasks. So no need to invoke onTerminated runnable method.

In our case as said above the `onAdvance()` will be called only once, when all registered parties are arrived at phaser current state. And since we have deregistered all arrived parties the onAdvance will return true and the phaser will be terminated

Here is a code sample to better understand:

```
public static void main(String[] args) {
		Phaser phaser = new Phaser(1) {
			@Override
			protected boolean onAdvance(int phase, int registeredParties) {
				boolean terminated = super.onAdvance(phase, registeredParties);
				if (terminated) {
					System.out.println("OnAdvance() invoked and phaser is terminated");
				} else {
					// WILL never happen
					throw new RuntimeException("OnAdvance() invoked and phaser is not terminated");
				}
				return terminated;
			}
		};
		String phaserStateTemplate = "\tRegistered parties: %s\n\tArrived parties still registered: %s\n\tUnArrived parties missing to advance to next phase: %s\n\tPhase: %s\n";
		System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
				phaser.getArrivedParties(), phaser.getUnarrivedParties(), phaser.getPhase()));

		// Following lines are to simulate the sse taks running before
		// shutdownNow method is invoked
		for (int i = 0; i < 2; i++) {
			System.out.println("Beginning of task " + i + ":");
			int currentPhase = phaser.register();
			System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
					phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));
			System.out.println("End of task " + i + ":");
			currentPhase = phaser.arriveAndDeregister();
			System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
					phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));
		}

		System.out.println("Shutdown invocation");
		// Following lines are to simulate the invocation of shutdownNow method
		int currentPhase = phaser.arriveAndDeregister();
		System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
				phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));

		System.out.println("Rejected task invocation");
		// Following lines are to simulate the sse taks running after
		// shutdownNow method is invoked
		currentPhase = phaser.arriveAndDeregister();
		System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
				phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));
	}
```

Reply:
@Thanks for the code example and it better explains. I added the new line in code example to try to explain what I meant that phaser.register() after phaser.arriveAndDeregister() doesn't increase the arrived party number from my local result:
registrationPhase = SseEventSourceScheduler.this.phaser.register();
(https://github.com/resteasy/Resteasy/pull/1677/files#diff-ef11a0ff038fff5fc91418bec89c4962R97)
```
  public static void main(String[] args) {
      Phaser phaser = new Phaser(1) {
          @Override
          protected boolean onAdvance(int phase, int registeredParties) {
              boolean terminated = super.onAdvance(phase, registeredParties);
              if (terminated) {
                  System.out.println("OnAdvance() invoked and phaser is terminated");
              } else {
                  // WILL never happen
                  throw new RuntimeException("OnAdvance() invoked and phaser is not terminated");
              }
              return terminated;
          }
      };
      String phaserStateTemplate = "\tRegistered parties: %s\n\tArrived parties still registered: %s\n\tUnArrived parties missing to advance to next phase: %s\n\tPhase: %s\n";
      System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
              phaser.getArrivedParties(), phaser.getUnarrivedParties(), phaser.getPhase()));

      // Following lines are to simulate the sse taks running before
      // shutdownNow method is invoked
      for (int i = 0; i < 2; i++) {
          System.out.println("Beginning of task " + i + ":");
          int currentPhase = phaser.register();
          System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
                  phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));
          System.out.println("End of task " + i + ":");
          currentPhase = phaser.arriveAndDeregister();
          
          System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
                  phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));
          //NEW ADDED & Below is the code to explain registrationPhase is always zero
          // int registrationPhase = SseEventSourceScheduler.this.phaser.register();
          System.out.println("*** phaser.register() after arriveAndDeregister *** " + phaser.register());
          System.out.println("*** phaser.getPhase() after arriveAndDeregister *** " + phaser.getPhase());
          phaser.arriveAndDeregister();
          
      }

      System.out.println("Shutdown invocation");
      // Following lines are to simulate the invocation of shutdownNow method
      int currentPhase = phaser.arriveAndDeregister();
      System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
              phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));

      System.out.println("Rejected task invocation");
      // Following lines are to simulate the sse taks running after
      // shutdownNow method is invoked
      currentPhase = phaser.arriveAndDeregister();
      System.out.println(String.format(phaserStateTemplate, phaser.getRegisteredParties(),
              phaser.getArrivedParties(), phaser.getUnarrivedParties(), currentPhase));
  }
```

> If onAdvance() is not called, it means that the phaser is not terminated and there are still running tasks. So no need to invoke onTerminated runnable method.

If I understand correctly, the onTerminated runnable method calls the all metehods registered onCompleteConsumers of SseEventSource in onAdance(). These runnables are all called after the EventSource is set to closed. Why we need to change this behavior ?

Reply:
@jimma, happy to ear that the code sample helps. I also get your point about the registration phase that will always be 0 before invocation of `shutdownNow()`. 
So I guess that we are done on this part.

Regarding the changes about calling onCompleteConsumers in `onAdvanceMethod()`:

I could have let the code as it was and do something like:

  ```
 private void internalClose()
    {
       if (state.getAndSet(State.CLOSED) == State.CLOSED)
       {
          return;
       }
      ...
      sseEventSourceScheduler.shutdownNow();
      onCompleteConsumers.forEach(Runnable::run);    
   }
```

But, IMO, there is a problem  with this approach.
To my mind, the `onCompleteConsumers.forEach(Runnable::run);`  is supposed to be invoked once no SSE task is running anymore. In other words `onCompleteConsumers.forEach(Runnable::run);` should be invoked once `SseEventSourceScheduler.awaitTermination(long timeout, TimeUnit unit)` returns `true`.
In code above, invocation of `sseEventSourceScheduler.shutdownNow()` does not guarantee for sure that there is no sse task running anymore. So we should not invoke `onCompleteConsumers.forEach(Runnable::run); ` here.
Instead if done in `onAdvance()` once `super.onAdvance()` return `true`,  we can say for sure that no sse task is/will be running anymore.
WDYT?

Reply:
>@jimma, happy to ear that the code sample helps. I also get your point about the registration phase that will always be 0 before invocation of shutdownNow().
So I guess that we are done on this part.

I am happy too that I finally make me understood. :-)   Let me give my more thoughts about this topic : 
 If phase that will always be 0 before invocation of shutdownNow(), does that mean this line can't be reached in any case :
``` 
 if (registrationPhase != 0) {
   return;
}
```
considering there is already closed check before this  and only shutdownNow can change registrationPhase to a negative value
```
     if (SseEventSourceScheduler.this.closed.get())
               {
                  return;
               }
```
From spec, onCompleteConsumers.forEach(Runnable::run)  should be invoked when there is no events received .  
I think we call it after setting closed state should be enough.   
https://github.com/eclipse-ee4j/jaxrs-api/blob/master/jaxrs-api/src/main/java/javax/ws/rs/sse/SseEventSource.java#L191

Reply:
Let's consider 2 cases with two threads T1 and T2:

- 1st case: 
T1  invokes `SseEventSourceScheduler.shutdownDown()` in one go without being interrupted by the underlying system and exits.
T2 (thread that will run the sse task) performs the closed check as you said and then exits since closed==true.
So in this case, you're right the check on `registrationPhase` is useless.

- 2nd case:
T1 (thread that will run the sse task) performs the closed check. Since `SseEventSourceScheduler.shutdownDown()`  has not been invoked yet, closed=false and T1 is allowed to continue. At the discretion of the underlying system where the java code is running, T1 is paused juste before running line (https://github.com/resteasy/Resteasy/pull/1677/files#diff-ef11a0ff038fff5fc91418bec89c4962R97) so that other threads can run.
T2  invokes `SseEventSourceScheduler.shutdownDown()` in one go without being interrupted by the underlying system and exits.
T1 is 'waken up' by the underlying system and line https://github.com/resteasy/Resteasy/pull/1677/files#diff-ef11a0ff038fff5fc91418bec89c4962R97 is executed.  At this moment, without any check on registration phase T1 would have been allowed to go ahead even if the phaser is terminated => this is not what we want.
So thank to the check on `registrationPhase` T1 will exit since `registrationPhase<0` because phaser is terminated.

So, yes, the second check is useful as seen above in case 2.

>I think we call it after setting closed state should be enough.

I will make the change. Thanks

Reply:
>2nd case:
T1 (thread that will run the sse task) performs the closed check. Since SseEventSourceScheduler.shutdownDown() has not been invoked yet, closed=false and T1 is allowed to continue. At the discretion of the underlying system where the java code is running, T1 is paused juste before running line (https://github.com/resteasy/Resteasy/pull/1677/files#diff-ef11a0ff038fff5fc91418bec89c4962R97) so that other threads can run.
T2 invokes SseEventSourceScheduler.shutdownDown() in one go without being interrupted by the underlying system and exits.
T1 is 'waken up' by the underlying system and line https://github.com/resteasy/Resteasy/pull/1677/files#diff-ef11a0ff038fff5fc91418bec89c4962R97 is executed. At this moment, without any check on registration phase T1 would have been allowed to go ahead even if the phaser is terminated => this is not what we want.
So thank to the check on registrationPhase T1 will exit since registrationPhase<0 because phaser is terminated.

Can you explain why T1 is paused and resumed ?

Reply:
Because, for any reason depending on the undelrying system (a higher priority thread becomes runnable, the time allotment reached for a given thread...).  a thread execution can be paused (and then resume) at any time during its execution for thread scheduling purpose in order to let other threads run.

Reply:
If you're not confortable with pause/resume, let's say that T1 and T2 are racing together.
T1 pass the check close and just before executing line https://github.com/resteasy/Resteasy/pull/1677/files#diff-ef11a0ff038fff5fc91418bec89c4962R97, T2 at the same time invokes `phaser.arriveAndDeregister()` in `SseEventSourceScheduler.shutdownDown()`.
T1 executes line https://github.com/resteasy/Resteasy/pull/1677/files#diff-ef11a0ff038fff5fc91418bec89c4962R97. At this moment, without any check on registration phase T1 would have been allowed to go ahead even if the phaser is terminated => this is not what we want.
So thank to the check on registrationPhase T1 will exit since registrationPhase<0 because phaser is terminated.

Globaly when multiple threads are racing, you cannot know in advance the way thread execution will be scheduled by the system (unless using specific mechanism such as mutual exclusion for instance) so we have to take all cases into account.

Reply:
@NicoNes  Thanks for clarification, and what you meant is cpu time/OS thread scheduling, not the java thread thing. From my understanding, Phaser or Phaser.register() is thread safe. Am I wrong ?

Reply:
>and what you meant is cpu time/OS thread scheduling, not the java thread thing

Yes exactly, that's it.

>From my understanding, Phaser or Phaser.register() is thread safe. Am I wrong ?

Yes you're rigth, T1 and T2 can share and use the same phaser instance safely

Reply:
@jimma I've just made the change about completion listeners notification.